packages/graphiql-react/src/editor/components/increments-editors.tsx

```diff
@@ -0,0 +1,141 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+
+import { ChevronDownIcon, ChevronUpIcon } from '../../icons';
+import { UnStyledButton } from '../../ui';
+import {
+  commonKeys,
+  DEFAULT_EDITOR_THEME,
+  DEFAULT_KEY_MAP,
+  importCodeMirror,
+} from '../common';
+import { useSynchronizeOption } from '../hooks';
+import { IncrementalPayload } from '../tabs';
+import { CodeMirrorEditor, CommonEditorProps } from '../types';
+
+import '../style/codemirror.css';
+import '../style/fold.css';
+import '../style/lint.css';
+import '../style/hint.css';
+import '../style/info.css';
+import '../style/jump.css';
+import '../style/auto-insertion.css';
+import '../style/editor.css';
+import '../style/increments-editors.css';
+
+type UseIncrementsEditorArgs = CommonEditorProps & {
+  increment: IncrementalPayload;
+};
+
+function useIncrementsEditor({
+  editorTheme = DEFAULT_EDITOR_THEME,
+  keyMap = DEFAULT_KEY_MAP,
+  increment,
+}: UseIncrementsEditorArgs) {
+  const [editor, setEditor] = useState<CodeMirrorEditor | null>(null);
+
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    let isActive = true;
+    void importCodeMirror(
+      [
+        import('codemirror/addon/fold/foldgutter'),
+        import('codemirror/addon/fold/brace-fold'),
+        import('codemirror/addon/dialog/dialog'),
+        import('codemirror/addon/search/search'),
+        import('codemirror/addon/search/searchcursor'),
+        import('codemirror/addon/search/jump-to-line'),
+        // @ts-expect-error
+        import('codemirror/keymap/sublime'),
+        import('codemirror-graphql/esm/results/mode'),
+        import('codemirror-graphql/esm/utils/info-addon'),
+      ],
+      { useCommonAddons: false },
+    ).then(CodeMirror => {
+      // Don't continue if the effect has already been cleaned up
+      if (!isActive) {
+        return;
+      }
+
+      const container = ref.current;
+      if (!container) {
+        return;
+      }
+
+      const newEditor = CodeMirror(container, {
+        value: JSON.stringify(increment.payload, null, 2),
+        lineWrapping: true,
+        readOnly: true,
+        theme: editorTheme,
+        mode: 'graphql-results',
+        foldGutter: true,
+        gutters: ['CodeMirror-foldgutter'],
+        // @ts-expect-error
+        info: true,
+        extraKeys: commonKeys,
+      });
+
+      setEditor(newEditor);
+    });
+
+    return () => {
+      isActive = false;
+    };
+  }, [editorTheme]);
+
+  useSynchronizeOption(editor, 'keyMap', keyMap);
+
+  return ref;
+}
+
+function IncrementEditor(
+  props: UseIncrementsEditorArgs & { isInitial: boolean },
+) {
+  const [isOpen, setIsOpen] = useState(false);
+  const incrementEditor = useIncrementsEditor(props);
+
+  const toggleEditor = useCallback(() => setIsOpen(current => !current), []);
+
+  return (
+    <div
+      className="graphiql-increment-editor"
+      style={isOpen ? { height: '30vh' } : {}}
```

Comment:
For the editors to show properly I needed to give the wrapper a fixed height, which is not ideal. But I don't see a better way right now other than starting to do JS magic to calculate the container height on-runtime.